File: Two Sum - Difference equals to target/Javascript/solution_3.js
// https://www.lintcode.com/problem/610/

function twoSum7(nums, target) {
  let len = nums.length;

  for (let i = 0; i < len; i++) {
    let num1 = nums[i];
    let num2 = num1 + Math.abs(target);
    if (binary_search(nums, num2, i + 1) !== -1) {
      return [num1, num2];
    }
  }

  return [];
}

function binary_search(nums, target, start) {
  // 对数组进行空值检查
  let nums_len = nums.length;
  if (!nums || nums_len === 0) {
    return -1;
  }

  let left = start;
  let right = nums.length - 1;

  while (left + 1 < right) {
    let mid = Math.floor((left + right) / 2);
    if (nums[mid] === target) {
      right = mid;
    } else if (nums[mid] > target) {
      right = mid;
    } else {
      left = mid;
    }
  }

  if (nums[left] === target) {
    return left;
  }

  if (nums[right] === target) {
    return right;
  }

  return -1;
}

let nums = [0, 1, 2, 2];
let target = 0;
console.log(twoSum7(nums, target));

File: Two Sum - Difference equals to target/Rust/src/solution_3.rs
// https://www.lintcode.com/problem/610/

pub fn two_sum7(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let nums_len = nums.len();
    let mut result = vec![];

    for i in 0..nums_len {
        let num1 = nums[i];
        let num2 = num1 + target.abs();
        if binary_search(&nums, num2, (i + 1) as i32) != -1 {
            return vec![num1, num2];
        }
    }

    result
}

pub fn binary_search(nums: &Vec<i32>, target: i32, start: i32) -> i32 {
    // 对数组进行空值检查
    let nums_len = nums.len();
    if nums_len == 0 {
        return -1;
    }

    let mut start = start;
    let mut end = (nums_len - 1) as i32;

    while start + 1 < end {
        let mid = start + (end - start) / 2;
        let mid_val = nums[mid as usize];

        if mid_val == target {
            end = mid;
        } else if mid_val > target {
            end = mid
        } else {
            start = mid;
        }
    }

    if nums[start as usize] == target {
        return start;
    }

    if nums[end as usize] == target {
        return end;
    }

    return -1;
}